exporter/metric/prometheus/prometheus.go

```diff
@@ -0,0 +1,184 @@
+	"go.opentelemetry.io/otel/sdk/export/metric/aggregator"
+)
+
+type metricKey struct {
+	desc    *export.Descriptor
+	encoded string
+}
+
+// Exporter is an implementation of metric.Exporter that sends metrics to
+// Prometheus.
+type Exporter struct {
+	handler http.Handler
+
+	registerer prometheus.Registerer
+	gatherer   prometheus.Gatherer
+
+	counters           counters
+	gauges             gauges
+	histograms         histograms
+	summaries          summaries
+	measureAggregation MeasureAggregation
+}
+
+var _ export.Exporter = &Exporter{}
+var _ http.Handler = &Exporter{}
+
+// Options is a set of options for the tally reporter.
+type Options struct {
+	// Registry is the prometheus registry that will be used as the default Registerer and
+	// Gatherer if these are not specified.
+	//
+	// If not set a new empty Registry is created.
+	Registry *prometheus.Registry
+
+	// Registerer is the prometheus registerer to register
+	// metrics with.
+	//
+	// If not specified the Registry will be used as default.
+	Registerer prometheus.Registerer
+
+	// Gatherer is the prometheus gatherer to gather
+	// metrics with.
+	//
+	// If not specified the Registry will be used as default.
+	Gatherer prometheus.Gatherer
+
+	// DefaultHistogramBuckets is the default histogram buckets
+	// to use. Use nil to specify the system-default histogram buckets.
+	DefaultHistogramBuckets []float64
+
+	// DefaultSummaryObjectives is the default summary objectives
+	// to use. Use nil to specify the system-default summary objectives.
+	DefaultSummaryObjectives map[float64]float64
+
+	// MeasureAggregation defines how metric.Measure are exported.
+	// Possible values are 'Histogram' or 'Summary'.
+	// The default export representation for measures is Histograms.
+	MeasureAggregation MeasureAggregation
+}
+
+type MeasureAggregation int
+
+const (
+	Histogram MeasureAggregation = iota
+	Summary
+)
+
+// NewExporter returns a new prometheus exporter for prometheus metrics.
+func NewExporter(opts Options) (*Exporter, error) {
+	if opts.Registry == nil {
+		opts.Registry = prometheus.NewRegistry()
+	}
+
+	if opts.Registerer == nil {
+		opts.Registerer = opts.Registry
+	}
+
+	if opts.Gatherer == nil {
+		opts.Gatherer = opts.Registry
+	}
+
+	return &Exporter{
+		registerer:         opts.Registerer,
+		gatherer:           opts.Gatherer,
+		handler:            promhttp.HandlerFor(opts.Gatherer, promhttp.HandlerOpts{}),
+		measureAggregation: opts.MeasureAggregation,
+
+		counters:   newCounters(opts.Registerer),
+		gauges:     newGauges(opts.Registerer),
+		histograms: newHistograms(opts.Registerer, opts.DefaultHistogramBuckets),
+		summaries:  newSummaries(opts.Registerer, opts.DefaultSummaryObjectives),
+	}, nil
+}
+
+// Export exports the provide metric record to prometheus.
+func (e *Exporter) Export(_ context.Context, checkpointSet export.CheckpointSet) error {
+	var forEachError error
+	checkpointSet.ForEach(func(record export.Record) {
```

Comment:
isn't a guard required if `forEachError` is nil or not ?

Reply:
We've being following the thought that if a single record fails to be exported that should not prevent the rest of the records to be processed.

There is still some questions we need to figure out about how to handle errors better on this `ForEach` method. One of the suggestions is to change its argument from `func(export.Record)` to `func (export.Record) error` to have a more uniform and simpler error handling on all exporters. But that is an open question that I think should be taken on a separate PR since it would touch not only this exporter but others too.